libevmasm/Assembly.cpp

```diff
@@ -1281,6 +1282,21 @@ LinkerObject const& Assembly::assembleLegacy() const
 	uint8_t tagPush = static_cast<uint8_t>(pushInstruction(bytesPerTag));
 	uint8_t dataRefPush = static_cast<uint8_t>(pushInstruction(bytesPerDataRef));
 
+	LinkerObject::CodeSectionLocation codeSectionLocation;
+	codeSectionLocation.start = 0;
+	size_t assemblyItemIndex = 0;
+	auto assembleInstruction = [&](auto&& _addInstruction) {
+		size_t start = ret.bytecode.size();
+		_addInstruction();
+		size_t end = ret.bytecode.size();
+		codeSectionLocation.instructionLocations.emplace_back(
+			LinkerObject::InstructionLocation{
+				.start = start,
+				.end = end,
+				.assemblyItemIndex = assemblyItemIndex
+			}
+		);
+	};
```

Comment:
Regarding the helper, I had something like this in mind:

```suggestion
	auto assembleInstruction = [](LinkerObject const& _linkerObject, Instruction _opcode, bytes const& _immediates, size_t assemblyItemIndex) {
		_linkerObject.bytecode += static_cast<uint8_t>(_opcode);
		_linkerObject.bytecode += _immediates;
		_linkerObject.codeSectionLocations[0].instructionLocations.push_back(
			LinkerObject::InstructionLocation{
				.start = ret.bytecode.size(),
				.end = ret.bytecode.size() 1 + _immediates.size(),
				.assemblyItemIndex = assemblyItemIndex
			}
		);
		_linkerObject.codeSectionLocations[0].end += 1 + _immediates.size();
	};
```

I.e. not depending on any external state, just arguments and clearly separating the bytecode from other stuff.

I'd not even make it a lambda - we'll need it in `assembleEOF()` later, so it should be a normal function, maybe even a class member.

Reply:
I'd even consider turning it into a class doing what it does currently as side-effect :-). But I'd say we can also wait until we inevitably port it to ``assemblyEOF`` to nicen it up.

Reply:
Well, it would be good to at least make it take bytecode, not a callback.

But yeah, it's not like it's a blocker here. It would be very nice to have, but I approved already because at this point these are not things critical to the functionality.

Reply:
But yeah, for the record: this is how I'd have done this: https://github.com/ethereum/solidity/compare/develop...ethdebug_instructions_and_source_ranges_refactor
Which includes doing it for EOF (which is just two lines in that version).
Just to keep anything ethdebug-related non-critical separate from the rest as much as possible (I'd move the helper classes out of the file).
The only uglyiness there is the ``AssignImmutable`` case, which unfortunately is irregular compared to all other cases, but it concentrates that uglyness there. But yeah, we can see if we do something like that or some other solution after the release, maybe there's something even better.